client/src/main/java/org/threadly/litesockets/client/http/HTTPClient.java

```diff
@@ -0,0 +1,406 @@
+    return requestAsync(ha, request, EMPTY_BUFFER);
+  }
+  
+  public ListenableFuture<HTTPResponseData> requestAsync(final HTTPAddress ha, final HTTPRequest request, final ByteBuffer body) {
+    return requestAsync(ha, request, body, TimeUnit.MILLISECONDS, defaultTimeout);
+  }
+  
+  public ListenableFuture<HTTPResponseData> requestAsync(final HTTPAddress ha, final HTTPRequest request, 
+      final ByteBuffer body, final TimeUnit tu, final long time) {
+    HTTPRequestWrapper hrw = new HTTPRequestWrapper(request, ha, body, tu.toMillis(time));
+    final ListenableFuture<HTTPResponseData> lf = hrw.slf;
+    queue.add(hrw);
+    if(ntse != null) {
+      ntse.wakeup();
+      if(isRunning.compareAndSet(false, true)) {
+        ssi.execute(new RunSocket());
+      }
+    } else {
+      processQueue();
+    }
+    return lf;
+  }
+  
+  private void processQueue() {
+    //This should be done after we do a .select on the ntse to check for more jobs before it exits.
+    while(maxConcurrent > inProcess.size()  && !queue.isEmpty()) {
+      HTTPRequestWrapper hrw = queue.poll();
+      //This runs concurrently if using a threaded SocketExecuter, so we have to null check before we add.
+      if(hrw != null) {
+        try {
+          hrw.client = getTCPClient(hrw.ha);
+          inProcess.put(hrw.client, hrw);
+          startWrite(hrw);
+        } catch (IOException e) {
+          hrw.slf.setFailure(e);
+        }
+      }
+    }
+  }
+
+  /**
+   * Stops the HttpClient.  Any requests currently pending will not be ran.  If this is not called
+   * and HTTPClient looses all references it will be GCed correctly. 
+   */
+  public void stop() {
+    if(ntse != null) {
+      ntse.stopIfRunning();
+      ntse.wakeup();
+      ntse.wakeup();
+    }
+    if(sts != null) {
+      sts.shutdownNow();
+    }
+  }
+  
+  private TCPClient getTCPClient(final HTTPAddress ha) throws IOException {
+    LinkedList<TCPClient> ll = sockets.get(ha);
+    TCPClient tc = null;
+    if(ll != null) {
+      synchronized(ll) {
+        while(ll.size() > 0 && tc == null) {
+          if(ll.peek().isClosed()) {
+            ll.pop();
+          } else {
+            tc = ll.pop();
+          }
+        }
+        if(ll.size() == 0) {
+          sockets.remove(ha);
+        }
+      }
+    }
+    if(tc == null) {
+      tc = sei.createTCPClient(ha.getHost(), ha.getPort());
+      if(ha.getdoSSL()) {
+        SSLEngine sse = SSLUtils.OPEN_SSL_CTX.createSSLEngine(ha.getHost(), ha.getPort());
+        sse.setUseClientMode(true);
+        tc.setSSLEngine(sse);
+        tc.startSSL();
+      }
+      tc.setReader(mcp);
+      tc.addCloseListener(mcp);
+      tc.connect();
+    }
+    return tc;
+  }
+  
+  private void addBackTCPClient(final HTTPAddress ha, final TCPClient client) {
+    if(!client.isClosed()) {
+      LinkedList<TCPClient> ll = sockets.get(ha);  
+      if(ll == null) {
+        sockets.put(ha, new LinkedList<TCPClient>());
```

Comment:
It seems based off the usage patterns that an ArrayDeque may be a more efficient structure here?